File: python/trakr_sdk/robot.py
import sys
from time import sleep
from enum import Enum

from .lib import trakr_python_sdk as sdk

from . import config_types as QuadDataTypes
from . import data_types as AlliedDataTypes

from .data_types import NDOF

class RobotMode(Enum):
    eLowLevel = 0
    eHighLevel = 1

class Robot:
    def __init__(self, ip : str, port : int, mode : int = RobotMode.eHighLevel):
        self.net_ = sdk.RobotNetwork()
        self.plan_ = sdk.AlliedPlan()
        self.state_ = sdk.AlliedState()

        self.setConfig_ : bool = False
        
        self.configStatus_ : int = 0

        self.ip = ip
        self.port = port
        self.mode = mode

    def setup(self, config : QuadDataTypes.CONFIG_SET, plan : AlliedDataTypes.Plan) -> bool:
        if(not self.net_.setup()):
            print("[ROBOT] Failed to setup Network")
            return False
        
        self._setConfigToPacket(config)
        self._setPlanToPacket(plan)
        self.plan_.config.status = 0

        if(not self.net_.setData(self.plan_)):
            print("[ROBOT] Failed to set data")
            return False
        
        if(not self.net_.connect(self.ip, self.port)):
            print("[ROBOT] Failed to connect to Network")
            return False
        
        print("[ROBOT] Connected!")

        self._bringUp()

        return True

    def run(self) -> bool:
        if(not self.net_.isConnected()):
            print("[ROBOT] Socket disconnected")
            return False

        if(self.net_.readData()):
            if(not self.net_.getData(self.state_)):
                print("[ROBOT] Unable to get data from Network")
                return False
            
            # Reset once acknowledged by server
            if(self.state_.config.status > 0):
                self.plan_.config.status = 0
                if(self.state_.config.status > 1):
                    self.plan_.config.status = -1
                self.configStatus_ = int(self.state_.config.status)
                self.state_.config.status = 0
            else:
                if(self.plan_.config.status < 0):
                    self.plan_.config.status = 0

        if(not self.net_.setData(self.plan_)):
            print("[ROBOT] Unable to set data to network")
            return False
        
        if(not self.net_.writeData()):
            print("[ROBOT] Failed to send data")
            return False
        
        return True
    
    def _bringUp(self) -> bool:
        print("[ROBOT] Waiting for Robot to BringUp!")

        self.state_.config.status = -1

        config_ = QuadDataTypes.CONFIG_SET()

        ctr = 0

        while((self.state_.config.status < 0) or (ctr < 10)):
            ctr += 1

            if(not self.run()):
                print("[ROBOT] Failed in BringUp")
                return False
            
            if(ctr < 10):
                self.getConfigStatus()
                sleep(0.01)
                continue

            # Some internal state machine to check status
            if(self.state_.config.status == -2):
                self.getConfig(config_)
                self.setConfig(config_)
                self.plan_.config.status = 1
            else:
                self.plan_.config.status = 0

            sleep(0.01)

        self.getConfigStatus()
        self.plan_.config.status = 0

        self._setMode(config_)

        print("[ROBOT] Robot BringUp Completed!")
        return True
    
    def _setMode(self, config : QuadDataTypes.CONFIG_SET):
        if(self.mode != RobotMode.eLowLevel):
            ctr = 0
            # Fetching latest data for some time (1s)
            while(self.isAlive() and (ctr < 50)):
                ctr += 1

                if(not self.run()):
                    print("[ROBOT] Failed in SetMode")
                    return False
                
                self.getConfigStatus()
                self.getConfig(config)

                sleep(0.02)
            return True
        
        ctr = 0
        # Fetching latest data for some time (2s)
        while(self.isAlive() and (ctr < 100)):
            ctr += 1

            if(not self.run()):
                print("[ROBOT] Failed in SetMode")
                return False
            
            self.getConfigStatus()
            self.getConfig(config)

            sleep(0.02)

        # Setting config for low-level mode
        while(self.isAlive()):
            if(not self.run()):
                print("[ROBOT] Failed in SetMode")
                return False

            if(self.getConfigStatus() == 2):
                break

            self.getConfig(config)
            config.motion.planner = QuadDataTypes.MotionDataTypes.TaskTypes.eMotion
            config.motion.strategy.type = QuadDataTypes.MotionDataTypes.MotionModes.eDeveloperMode
            self.setConfig(config)

            sleep(0.01)

        ctr = 0
        # Fetching latest data for some time (2s)
        while(self.isAlive() and (ctr < 125)):
            ctr += 1

            if(not self.run()):
                print("[ROBOT] Failed in SetMode")
                return False
            
            self.getConfigStatus()
            self.getConfig(config)

            sleep(0.02)
        print("After Loop")
        print("[ROBOT] Switched to LowLevel/Developer Mode")

        return True

    def isAlive(self) -> bool:
        return self.net_.isConnected()
    
    def setData(self, plan : AlliedDataTypes.Plan) -> bool:
        self._setPlanToPacket(plan)
        return True

    def getData(self, state : AlliedDataTypes.State) -> bool:
        self._getStateFromPacket(state)
        self.state_.config.status = 0
        return True

    def setConfig(self, config : QuadDataTypes.CONFIG_SET) -> bool:
        self._setConfigToPacket(config)
        if(self.plan_.config.status >= 0):
            self.plan_.config.status = 1
        return True

    def getConfig(self, config : QuadDataTypes.CONFIG_SET) -> bool:
        self._getConfigFromPacket(config)
        return True

    def getConfigStatus(self) -> int:
        ret = int(self.configStatus_)
        self.configStatus_ = 0
        return ret

    def _setPlanToPacket(self, plan : AlliedDataTypes.Plan):
        self.plan_.plan.joint.pos = plan.joint.pos
        self.plan_.plan.joint.vel = plan.joint.vel
        self.plan_.plan.joint.tor = plan.joint.tor
        self.plan_.plan.joint.kp = plan.joint.kp
        self.plan_.plan.joint.kd = plan.joint.kd

        self.plan_.plan.torso.pos = plan.torso.pos
        self.plan_.plan.torso.vel = plan.torso.vel


    def _setConfigToPacket(self, config : QuadDataTypes.CONFIG_SET):
        self.plan_.config.master.shutdown = config.master.shutdown
        self.plan_.config.master.exit = config.master.exit
        self.plan_.config.master.killed = config.master.killed

        self.plan_.config.motion.planner = config.motion.planner.value
        self.plan_.config.motion.sequence.type = config.motion.sequence.type.value
        self.plan_.config.motion.sequence.seq = 0
        self.plan_.config.motion.strategy.type = config.motion.strategy.type.value
        if(config.motion.strategy.type == QuadDataTypes.MotionDataTypes.MotionModes.eClassicalMode):
            self.plan_.config.motion.strategy.seq = config.motion.strategy.seq.value
        elif(config.motion.strategy.type == QuadDataTypes.MotionDataTypes.MotionModes.eAIMode):
            self.plan_.config.motion.strategy.seq = config.motion.strategy.seq.value
        else:
            self.plan_.config.motion.strategy.seq = 0

        self.plan_.config.safety.llsafety.enabled = config.safety.llsafety.enabled
        self.plan_.config.safety.llsafety.jpos = config.safety.llsafety.check_jpos
        self.plan_.config.safety.llsafety.jvel = config.safety.llsafety.check_jvel
        self.plan_.config.safety.llsafety.jtor = config.safety.llsafety.check_jtor
        self.plan_.config.safety.llsafety.dtor = config.safety.llsafety.check_dtor

        self.plan_.config.safety.llsafety.lim_jpos_min = config.safety.llsafety.lim_jpos_min
        self.plan_.config.safety.llsafety.lim_jpos_max = config.safety.llsafety.lim_jpos_max
        self.plan_.config.safety.llsafety.lim_jvel = config.safety.llsafety.lim_jvel
        self.plan_.config.safety.llsafety.lim_jtor = config.safety.llsafety.lim_jtor
        self.plan_.config.safety.llsafety.lim_dtor = config.safety.llsafety.lim_dtor

        self.plan_.config.safety.llheartbeat.enabled = config.safety.llheartbeat.enabled
        self.plan_.config.safety.llheartbeat.init_disable_time = 1
        self.plan_.config.safety.llheartbeat.enable_CAN = config.safety.llheartbeat.check_joints
        self.plan_.config.safety.llheartbeat.CAN_timeout = config.safety.llheartbeat.joints_timeout
        self.plan_.config.safety.llheartbeat.observe_jpos = False
        self.plan_.config.safety.llheartbeat.observe_jvel = True
        self.plan_.config.safety.llheartbeat.observe_jtor = True
        self.plan_.config.safety.llheartbeat.enable_IMU = config.safety.llheartbeat.check_IMU
        self.plan_.config.safety.llheartbeat.IMU_timeout = config.safety.llheartbeat.IMU_timeout
        self.plan_.config.safety.llheartbeat.observe_accel = True
        self.plan_.config.safety.llheartbeat.observe_gyro = True

        self.plan_.config.safety.torso.enabled = config.safety.torso.enabled
        self.plan_.config.safety.torso.observe_roll = config.safety.torso.check_roll
        self.plan_.config.safety.torso.roll_limit = config.safety.torso.roll_limit
        self.plan_.config.safety.torso.observe_pitch = config.safety.torso.check_pitch
        self.plan_.config.safety.torso.pitch_limit = config.safety.torso.pitch_limit
        self.plan_.config.safety.torso.observe_gyro = config.safety.torso.check_gyro
        self.plan_.config.safety.torso.gyro_limit = config.safety.torso.gyro_limit


    def _getStateFromPacket(self, state : AlliedDataTypes.State):
        for i in range(NDOF):
            state.joint.pos[i] = self.state_.state.joint.pos[i]
            state.joint.vel[i] = self.state_.state.joint.vel[i]
            state.joint.tor[i] = self.state_.state.joint.tor[i]
            state.joint.kp[i] = self.state_.state.joint.kp[i]
            state.joint.kd[i] = self.state_.state.joint.kd[i]
        state.joint.timestamp = self.state_.state.joint.timestamp

        for i in range(6):
            state.torso.pos[i] = self.state_.state.torso.pos[i]
            state.torso.vel[i] = self.state_.state.torso.vel[i]
        state.torso.timestamp = self.state_.state.torso.timestamp

        for i in range(3):
            state.imu.acc[i] = self.state_.state.imu.acc[i]
            state.imu.mag[i] = self.state_.state.imu.mag[i]
            state.imu.gyro[i] = self.state_.state.imu.gyro[i]
            state.imu.euler[i] = self.state_.state.imu.euler[i]
        state.imu.timestamp = self.state_.state.imu.timestamp

    def _getConfigFromPacket(self, config : QuadDataTypes.CONFIG_SET):
        config.master.shutdown = self.state_.config.master.shutdown
        config.master.exit = self.state_.config.master.exit
        config.master.killed = self.state_.config.master.killed

        config.motion.planner = QuadDataTypes.MotionDataTypes.TaskTypes(self.state_.config.motion.planner)
        config.motion.sequence.type = QuadDataTypes.MotionDataTypes.GestureTypes(self.state_.config.motion.sequence.type)
        config.motion.sequence.seq = 0
        config.motion.strategy.type = QuadDataTypes.MotionDataTypes.MotionModes(self.state_.config.motion.strategy.type)
        try:
            config.motion.strategy.seq = QuadDataTypes.MotionDataTypes.MotionGaits(self.state_.config.motion.strategy.seq)
        except: 
            config.motion.strategy.seq = QuadDataTypes.MotionDataTypes.MotionGaits(0)

        config.safety.llsafety.enabled = self.state_.config.safety.llsafety.enabled
        config.safety.llsafety.check_jpos = self.state_.config.safety.llsafety.jpos
        config.safety.llsafety.check_jvel = self.state_.config.safety.llsafety.jvel
        config.safety.llsafety.check_jtor = self.state_.config.safety.llsafety.jtor
        config.safety.llsafety.check_dtor = self.state_.config.safety.llsafety.dtor

        config.safety.llsafety.lim_jpos_min = self.state_.config.safety.llsafety.lim_jpos_min
        config.safety.llsafety.lim_jpos_max = self.state_.config.safety.llsafety.lim_jpos_max
        config.safety.llsafety.lim_jvel = self.state_.config.safety.llsafety.lim_jvel
        config.safety.llsafety.lim_jtor = self.state_.config.safety.llsafety.lim_jtor
        config.safety.llsafety.lim_dtor = self.state_.config.safety.llsafety.lim_dtor

        config.safety.llheartbeat.enabled = self.state_.config.safety.llheartbeat.enabled
        config.safety.llheartbeat.check_joints = self.state_.config.safety.llheartbeat.enable_CAN
        config.safety.llheartbeat.joints_timeout = self.state_.config.safety.llheartbeat.CAN_timeout
        config.safety.llheartbeat.check_IMU = self.state_.config.safety.llheartbeat.enable_IMU
        config.safety.llheartbeat.IMU_timeout = self.state_.config.safety.llheartbeat.IMU_timeout

        config.safety.torso.enabled = self.state_.config.safety.torso.enabled
        config.safety.torso.check_roll = self.state_.config.safety.torso.observe_roll
        config.safety.torso.roll_limit = self.state_.config.safety.torso.roll_limit
        config.safety.torso.check_pitch = self.state_.config.safety.torso.observe_pitch
        config.safety.torso.pitch_limit = self.state_.config.safety.torso.pitch_limit
        config.safety.torso.check_gyro = self.state_.config.safety.torso.observe_gyro
        config.safety.torso.gyro_limit = self.state_.config.safety.torso.gyro_limit

File: cpp/trakr_cpp_sdk/src/robot.cpp
#include "robot.h"

/**
 * @brief Perform network setup and connection
 *
 */
bool Robot::setup(QuadDataTypes::CONFIG_SET &config, AlliedDataTypes::Plan &plan)
{
    if(!net_.setup())
    {
        return false;
    }

    dataAdaptor_(plan_, config);
    dataAdaptor_(plan_, plan);
    plan_.config.status_ = false;

    if(!net_.setData(plan_))
    {
        return false;
    }

    if(!net_.connect(ip_addr_, port_))
    {
        return false;
    }

    std::cout << "[ROBOT] Connected!" << std::endl;

    bringUp_();

    return true;
}

/**
 * @brief Communicate with robot (send and receive)
 *
 */
bool Robot::run()
{
    if(!net_.readData())
    {
        std::cout << "[ROBOT] Failed to read data" << std::endl;
        return false;
    }

    if(!net_.getData(state_))
    {
        return false;
    }

    // Reset once acknowledged by server
    if((state_.config.status_ > 0))
    {
        plan_.config.status_ = 0;
        config_status_ = state_.config.status_;
        state_.config.status_ = 0;
    }

    if(!net_.setData(plan_))
    {
        return false;
    }

    if(!net_.writeData())
    {
        std::cout << "[ROBOT] Failed to send data" << std::endl;
        return false;
    }

    return true;
}


/**
 * @brief Checking if robot bringUp is done
 *
 */
bool Robot::bringUp_()
{
    std::cout << "[ROBOT] Waiting for Robot to BringUp!" << std::endl;

    state_.config.status_ = -1;

    QuadDataTypes::CONFIG_SET config;

    int ctr = 0;

    while(state_.config.status_ < 0)
    {
        ctr++;

        if(!run())
        {
            std::cout << "[ROBOT] Failed in BringUp" << std::endl;
            return false;
        }

        if(ctr < 10)
        {
            usleep(100000);
            continue;
        }

        // Some internal state machine to check status
        if(state_.config.status_ == -2)
        {
            getConfig(config);
            setConfig(config);
            plan_.config.status_ = 1;
        }
        else
        {
            plan_.config.status_ = 0;
        }
        usleep(100000);
    }

    plan_.config.status_ = 0;

    std::cout << "[ROBOT] Robot BringUp Completed!" << std::endl;
    return true;
}

/**
 * @brief Set the allied data to be sent
 *
 */
bool Robot::setData(AlliedDataTypes::Plan &plan)
{
    dataAdaptor_(plan_, plan);
    return true;
}

/**
 * @brief Get the received allied data
 *
 */
bool Robot::getData(AlliedDataTypes::State &state)
{
    dataAdaptor_(state_, state);
    return true;
}

/**
 * @brief Set the robot config to be sent
 *
 */
bool Robot::setConfig(QuadDataTypes::CONFIG_SET &config)
{
    dataAdaptor_(plan_, config);
    plan_.config.status_ = 1;
    return true;
}

/**
 * @brief Get the latest config most probably being run by robot
 *
 */
bool Robot::getConfig(QuadDataTypes::CONFIG_SET &config)
{
    dataAdaptor_(state_, config);
    return true;
}

/**
 * @brief Perform network setup and connection
 *
 */
int Robot::getConfigStatus()
{
    int ret = config_status_;
    config_status_ = 0;
    return ret;
}

/**
 * @brief Set allied data into local buffer
 *
 */
void Robot::dataAdaptor_(ClientNetworkConfig::AlliedPlan &send, AlliedDataTypes::Plan &plan)
{
    for(int i = 0; i < NDOF; i++)
    {
        send.plan.joint.pos[i] = plan.joint.pos[i];
        send.plan.joint.vel[i] = plan.joint.vel[i];
        send.plan.joint.tor[i] = plan.joint.tor[i];
        send.plan.joint.kp[i] = plan.joint.kp[i];
        send.plan.joint.kd[i] = plan.joint.kd[i];
    }

    for(int i = 0; i < 6; i++)
    {
        send.plan.torso.pos[i] = plan.torso.pos[i];
        send.plan.torso.vel[i] = plan.torso.vel[i];
    }
}

/**
 * @brief Set config into local buffer
 *
 */
void Robot::dataAdaptor_(ClientNetworkConfig::AlliedPlan &send, QuadDataTypes::CONFIG_SET &config)
{
    send.config.master.shutdown = config.master.shutdown;
    send.config.master.exit = config.master.exit;
    send.config.master.killed = config.master.killed;

    send.config.motion.planner = static_cast<int>(config.motion.planner);
    send.config.motion.sequence.type = static_cast<int>(config.motion.sequence.type);
    send.config.motion.sequence.seq = static_cast<int>(config.motion.sequence.seq);
    send.config.motion.strategy.type = static_cast<int>(config.motion.strategy.type);
    switch (config.motion.strategy.type)
    {
        case MotionModes::eClassicalMode :
        send.config.motion.strategy.seq = static_cast<int>(config.motion.strategy.seq);
        break;

        default:
        send.config.motion.strategy.seq = 0;
        break;
    }

    send.config.safety.llsafety.enabled = config.safety.llsafety.enabled;
    send.config.safety.llsafety.jpos = config.safety.llsafety.jpos;
    send.config.safety.llsafety.jvel = config.safety.llsafety.jvel;
    send.config.safety.llsafety.jtor = config.safety.llsafety.jtor;
    send.config.safety.llsafety.dtor = config.safety.llsafety.dtor;

    for(int i = 0; i < NDOF; i++)
    {
        send.config.safety.llsafety.lim_jpos_min[i] = config.safety.llsafety.lim_jpos_min[i];
        send.config.safety.llsafety.lim_jpos_max[i] = config.safety.llsafety.lim_jpos_max[i];
        send.config.safety.llsafety.lim_jvel[i] = config.safety.llsafety.lim_jvel[i];
        send.config.safety.llsafety.lim_jtor[i] = config.safety.llsafety.lim_jtor[i];
        send.config.safety.llsafety.lim_dtor[i] = config.safety.llsafety.lim_dtor[i];
    }

    send.config.safety.llheartbeat.enabled = config.safety.llheartbeat.enabled;
    send.config.safety.llheartbeat.init_disable_time = config.safety.llheartbeat.init_disable_time;
    send.config.safety.llheartbeat.enable_CAN = config.safety.llheartbeat.enable_CAN;
    send.config.safety.llheartbeat.CAN_timeout = config.safety.llheartbeat.CAN_timeout;
    send.config.safety.llheartbeat.observe_jpos = config.safety.llheartbeat.observe_jpos;
    send.config.safety.llheartbeat.observe_jvel = config.safety.llheartbeat.observe_jvel;
    send.config.safety.llheartbeat.observe_jtor = config.safety.llheartbeat.observe_jtor;
    send.config.safety.llheartbeat.enable_IMU = config.safety.llheartbeat.enable_IMU;
    send.config.safety.llheartbeat.IMU_timeout = config.safety.llheartbeat.IMU_timeout;
    send.config.safety.llheartbeat.observe_accel = config.safety.llheartbeat.observe_accel;
    send.config.safety.llheartbeat.observe_gyro = config.safety.llheartbeat.observe_gyro;

    send.config.safety.torso.enabled = config.safety.torso.enabled;
    send.config.safety.torso.observe_roll = config.safety.torso.observe_roll;
    send.config.safety.torso.roll_limit = config.safety.torso.roll_limit;
    send.config.safety.torso.observe_pitch = config.safety.torso.observe_pitch;
    send.config.safety.torso.pitch_limit = config.safety.torso.pitch_limit;
    send.config.safety.torso.observe_gyro = config.safety.torso.observe_gyro;
    send.config.safety.torso.gyro_limit = config.safety.torso.gyro_limit;
}

/**
 * @brief Set state from local buffer
 *
 */
void Robot::dataAdaptor_(ClientNetworkConfig::AlliedState &recv, AlliedDataTypes::State &state)
{
    for(int i = 0; i < NDOF; i++)
    {
        state.joint.pos[i] = recv.state.joint.pos[i];
        state.joint.vel[i] = recv.state.joint.vel[i];
        state.joint.tor[i] = recv.state.joint.tor[i];
        state.joint.kp[i] = recv.state.joint.kp[i];
        state.joint.kd[i] = recv.state.joint.kd[i];
    }

    for(int i = 0; i < 6; i++)
    {
        state.torso.pos[i] = recv.state.torso.pos[i];
        state.torso.vel[i] = recv.state.torso.vel[i];
    }

    for(int i = 0; i < 3; i++)
    {
        state.imu.acc[i] = recv.state.imu.acc[i];
        state.imu.gyro[i] = recv.state.imu.gyro[i];
        state.imu.mag[i] = recv.state.imu.mag[i];
        state.imu.euler[i] = recv.state.imu.gyro[i];
    }
}

/**
 * @brief Set config from local buffer
 *
 */
void Robot::dataAdaptor_(ClientNetworkConfig::AlliedState &recv, QuadDataTypes::CONFIG_SET &config)
{
    config.master.shutdown = recv.config.master.shutdown;
    config.master.exit = recv.config.master.exit;
    config.master.killed = recv.config.master.killed;

    config.motion.planner = static_cast<MotionDataTypes::TaskTypes>(recv.config.motion.planner);
    config.motion.sequence.type = static_cast<GestureTypes>(recv.config.motion.sequence.type);
    config.motion.sequence.seq = static_cast<int>(recv.config.motion.sequence.seq);
    config.motion.strategy.type = static_cast<MotionModes>(recv.config.motion.strategy.type);
    config.motion.strategy.seq = static_cast<MotionGaits>(recv.config.motion.strategy.seq);

    config.safety.llsafety.enabled = recv.config.safety.llsafety.enabled;
    config.safety.llsafety.jpos = recv.config.safety.llsafety.jpos;
    config.safety.llsafety.jvel = recv.config.safety.llsafety.jvel;
    config.safety.llsafety.jtor = recv.config.safety.llsafety.jtor;
    config.safety.llsafety.dtor = recv.config.safety.llsafety.dtor;

    for(int i = 0; i < NDOF; i++)
    {
        config.safety.llsafety.lim_jpos_min[i] = recv.config.safety.llsafety.lim_jpos_min[i];
        config.safety.llsafety.lim_jpos_max[i] = recv.config.safety.llsafety.lim_jpos_max[i];
        config.safety.llsafety.lim_jvel[i] = recv.config.safety.llsafety.lim_jvel[i];
        config.safety.llsafety.lim_jtor[i] = recv.config.safety.llsafety.lim_jtor[i];
        config.safety.llsafety.lim_dtor[i] = recv.config.safety.llsafety.lim_dtor[i];
    }

    config.safety.llheartbeat.enabled = recv.config.safety.llheartbeat.enabled;
    config.safety.llheartbeat.init_disable_time = recv.config.safety.llheartbeat.init_disable_time;
    config.safety.llheartbeat.enable_CAN = recv.config.safety.llheartbeat.enable_CAN;
    config.safety.llheartbeat.CAN_timeout = recv.config.safety.llheartbeat.CAN_timeout;
    config.safety.llheartbeat.observe_jpos = recv.config.safety.llheartbeat.observe_jpos;
    config.safety.llheartbeat.observe_jvel = recv.config.safety.llheartbeat.observe_jvel;
    config.safety.llheartbeat.observe_jtor = recv.config.safety.llheartbeat.observe_jtor;
    config.safety.llheartbeat.enable_IMU = recv.config.safety.llheartbeat.enable_IMU;
    config.safety.llheartbeat.IMU_timeout = recv.config.safety.llheartbeat.IMU_timeout;
    config.safety.llheartbeat.observe_accel = recv.config.safety.llheartbeat.observe_accel;
    config.safety.llheartbeat.observe_gyro = recv.config.safety.llheartbeat.observe_gyro;

    config.safety.torso.enabled = recv.config.safety.torso.enabled;
    config.safety.torso.observe_roll = recv.config.safety.torso.observe_roll;
    config.safety.torso.roll_limit = recv.config.safety.torso.roll_limit;
    config.safety.torso.observe_pitch = recv.config.safety.torso.observe_pitch;
    config.safety.torso.pitch_limit = recv.config.safety.torso.pitch_limit;
    config.safety.torso.observe_gyro = recv.config.safety.torso.observe_gyro;
    config.safety.torso.gyro_limit = recv.config.safety.torso.gyro_limit;
}
